### ISSUE TITLE
Not working on Windows (Sublime build 3142)

### ISSUE THREAD

**Issue by mbrodala** [gh640/SublimePhpArrayConverter#2]
When triggering the command for syntax conversion nothing happens, only a message is shown in the status bar:

> PhpArrayConverter tokenizer failed: .

There is nothing special required to see this, the following very basic example shows the error:

```php
$foo = array();
```

**Comment by gh640** [comment]
Thank you for trying the package.

There are mainly 2 steps in this conversion process:

1. Tokenize a piece of code with `php` command.
2. Replace arrays in the code using the tokenized infomation from 1.

The error message `PhpArrayConverter tokenizer failed:` is shown when the step 1 failed. Normally the original error message is inserted between `tokenizer failed: ` and `.`. There seems to be no message...

If possible, try `php /path/to/the/PhpArrayConverter/tokenizer.php < [your php code]` and see the result. It may give us useful info for debugging.

Or, can you share more information where it doesn't work as much as possible? For example, I guess it may not yet work in Windows properly since I don't have a Windows machine at the moment.

Thanks.

**Comment by mbrodala** [comment]
Yes, I'm on Windows and in fact piping code to the `tokenizer.php` didn't work with Git Bash (`stdin is not a tty`) so I had to switch to `get_code_from_file()`. This works fine with `cmd.exe` though.

In any case the tokenizer seems to work fine. Here's the test file:

```php
<?php
$foo = array();
```

And the tokenizer output:

```json
{"tokens":[["T_OPEN_TAG","<?php\n",1],[320,"$foo",2],["T_WHITESPACE"," ",2],"=",["T_WHITESPACE"," ",
2],["T_ARRAY","array",2],"(",")",";",["T_WHITESPACE","\n",2]]}
```

**Comment by gh640** [comment]
Thank you for inspecting the piece. I see. The result looks fine. I understood that there's no problem with the php script for the tokenization. The problem seems to be in the python code which invokes the php script. 

The python file which runs the tokenizer is `PhpArrayConverter.py` in the package root. There's a class named `PhpTokenizer` in it. It runs the php tokenizer through `subprocess.Popen` in `run()` method. Currently the argument `shell` passed to the constructor `Popen()` is set to `True` on Windows and it may be problematic. Could you please try setting it to `False`? Actually the default value of `shell` is `False` and you just have to delete the line to set it to `False`. A sample patch is as following: 

```diff
diff --git a/PhpArrayConverter.py b/PhpArrayConverter.py
index 4e9e393..3f70711 100644
--- a/PhpArrayConverter.py
+++ b/PhpArrayConverter.py
@@ -117,7 +117,7 @@ class PhpTokenizer(object):
             process = Popen(
                 self.get_php_cmd(),
                 env=self.get_env(),
-                shell=sublime.platform() == 'windows',
+                # shell=sublime.platform() == 'windows',
                 stdin=PIPE,
                 stdout=PIPE,
                 stderr=PIPE
```

I'd like you to try running the `PhpArrayConverter` command again after commenting out or deleting the line. Thank you for your cooperation!

**Comment by mbrodala** [comment]
OK, this is a bit weird. I've cloned the repository into `Packages` as [described here](https://packagecontrol.io/docs/customizing_packages), made the suggested change and removed the `Installed Packages/PhpArrayConverter.sublime-package` file.

Now when I try to convert with the example given above, it works exactly one time. If I revert the change and try again the mentioned error appears again. Strangely Sublime (or PackageControl) automatically restores the original `Installed Packages/PhpArrayConverter.sublime-package` file after doing one conversion. That package contains the original (unmodified) code.

So unless you have an idea how to keep the unpackaged package, I can perform a final test only after a new release.

**Comment by gh640** [comment]
Thanks. I see.

> Strangely Sublime (or PackageControl) automatically restores the original Installed Packages/PhpArrayConverter.sublime-package file ...

I don't know much about it but AFAIK Sublime Text 3 seems to restore (fix) packages in `Installed Packages` directory automatically.

> by default, Package Control will automatically upgrade packages to the newest version. This will cause any edits to files to be overwritten.

https://packagecontrol.io/docs/customizing_packages#Editing_Unpacked_Files

So I'd like you to try uninstalling PhpArrayConverter with the command `Package Control: Remove Package` instead of removing `Installed Packages/PhpArrayConverter.sublime-package` directly. Confirm that the directory is removed, clone the repository into `Packages`, try the modification and test the command (array conversion) again.

I'll make a new release. Thank you.

**Comment by mbrodala** [comment]
OK, did that now and the fix seems to work reliably.

However, I noticed that now a console window shortly flashes up when converting. I assume that's exactly what the `shell` argument avoids.

**Comment by gh640** [comment]
> However, I noticed that now a console window shortly flashes up when converting. I assume that's exactly what the `shell` argument avoids.

Specifying `STARTUPINFO` on Windows when using `Popen` may prevent the window creation. I tried making a commit to fix it. Anyway, could you try the latest `master`?

https://github.com/gh640/SublimePhpArrayConverter/commit/5f68b29c879de2d19f98e77bb58641e75d41596a

**Comment by mbrodala** [comment]
Works like a charm, conversion keeps working on multiple runs and no consolewindow appears. 👍

**Comment by gh640** [comment]
I'm happy to hear that. I've pushed a new release (0.1.3) and it will be recognized to Package Control soon. Thanks!

**Comment by mbrodala** [comment]
Hm, it's weird. I've now removed the master clone in `Packages` and installed the new release regularly and the original error shows up again. Removing the package and using the latest master instead makes it working again however...

**Comment by gh640** [comment]
Thank you for sharing the issue. Maybe there's another cause on this. I'll tackle it and please wait a little. Please use `master` until then.

**Comment by gh640** [comment]
Sorry for the inconvenience. The problem might always happen when the package is downloaded with Package Control and the package is kept packed as `.sublime-package`.

So I've added an empty file `.no-sublime-package` under the root directory so that the package is unpacked after installation: https://packagecontrol.io/docs/customizing_packages

I've released a new version 0.1.4 and it's available through Package Control now. I'd like you to try it.

https://packagecontrol.io/packages/PhpArrayConverter

When switching from the latest master to the standard package, you might need to reboot Sublime Text 3 itself. Thank you.
